file/file_prefetch_buffer.cc

```diff
@@ -399,289 +376,285 @@ void FilePrefetchBuffer::ReadAheadSizeTuning(
-    PollAndUpdateBuffersIfNeeded(offset, length);
+  if (buf->async_read_in_progress_ &&
+      buf->IsOffsetInBufferWithAsyncProgress(offset)) {
+    PollIfNeeded(offset, length);
   }
-  second = curr_ ^ 1;
 
-  // If data is overlapping over two buffers, copy the data from curr_ and
-  // call ReadAsync on curr_.
-  if (!bufs_[curr_].async_read_in_progress_ && DoesBufferContainData(curr_) &&
-      IsOffsetInBuffer(offset, curr_) &&
-      (/*Data extends over curr_ buffer and second buffer either has data or in
+  BufferInfo* next_buf = bufs_[1];
+
+  // If data is overlapping over two buffers, copy the data from front and
+  // call ReadAsync on freed buffer.
+  if (!buf->async_read_in_progress_ && buf->DoesBufferContainData() &&
+      buf->IsOffsetInBuffer(offset) &&
+      (/*Data extends over two buffers and second buffer either has data or in
          process of population=*/
-       (offset + length > bufs_[second].offset_) &&
-       (bufs_[second].async_read_in_progress_ ||
-        DoesBufferContainData(second)))) {
-    // Allocate new buffer to third buffer;
-    bufs_[2].ClearBuffer();
-    bufs_[2].buffer_.Alignment(alignment);
-    bufs_[2].buffer_.AllocateNewBuffer(length);
-    bufs_[2].offset_ = offset;
-    copy_to_third_buffer = true;
-
-    CopyDataToBuffer(curr_, tmp_offset, tmp_length);
-
-    // Call async prefetching on curr_ since data has been consumed in curr_
-    // only if requested data lies within second buffer.
-    size_t second_size = bufs_[second].async_read_in_progress_
-                             ? bufs_[second].async_req_len_
-                             : bufs_[second].buffer_.CurrentSize();
-    uint64_t start_offset = bufs_[second].initial_end_offset_;
+       (offset + length > next_buf->offset_) &&
+       (next_buf->async_read_in_progress_ ||
+        next_buf->DoesBufferContainData()))) {
+    // Allocate new buffer to overlap_buf_.
+    overlap_buf_->ClearBuffer();
+    overlap_buf_->buffer_.Alignment(alignment);
+    overlap_buf_->buffer_.AllocateNewBuffer(length);
+    overlap_buf_->offset_ = offset;
+    copy_to_overlap_buffer = true;
+
+    size_t initial_buf_size = overlap_buf_->CurrentSize();
+    CopyDataToBuffer(buf, tmp_offset, tmp_length);
+    UpdateStats(
+        /*found_in_buffer=*/false,
+        overlap_buf_->CurrentSize() - initial_buf_size);
+
+    // Call async prefetching on freed buffer since data has been consumed
+    // only if requested data lies within next buffer.
+    size_t second_size = next_buf->async_read_in_progress_
+                             ? next_buf->async_req_len_
+                             : next_buf->CurrentSize();
+    uint64_t start_offset = next_buf->initial_end_offset_;
     // Second buffer might be out of bound if first buffer already prefetched
     // that data.
-    if (tmp_offset + tmp_length <= bufs_[second].offset_ + second_size) {
+
+    if (tmp_offset + tmp_length <= next_buf->offset_ + second_size) {
```

Comment:
What does it mean for `tmp_offset + tmp_length > next_buf->offset_ + second_size`? The read is not done but we're also not going to read more?

I also couldn't understand the comment: "Second buffer might be out of bound ...". What bounds are we referring to?

Reply:
Oh let me update the comment. I forgot to remove this comment.

Reply:
If requested bytes - tmp_offset + tmp_length are in next buffer, freed buffer can go for further prefetching.
If requested bytes are not in next buffer, next buffer has to go for sync call to get remaining requested bytes. In that it shouldn't go for async  prefetching as async prefetching calculates offset based on previous buffer end offset and previous buffer has to go for sync prefetching.

Reply:
It makes sense. I might think about how to write a loop that incorporates data from three or more async buffers, but haven't thought of anything to propose here yet.

Reply:
> think about how to write a loop that incorporates data from three or more async buffers

It's also useful to consolidate the handling for issuing more async readahead after freeing an outdated buffer. Right now I think the flow goes (1) free an outdated buffer (`ClearOutdatedData()`), (2) poll the front buffer, and then (3) allocate a new buffer and issue async readahead on it (`PrefetchRemBuffers()`). But if a buffer is consumed by `HandleOverlappingData()`, the new async readahead is issued before polling the next front buffer. The latter strategy has a little bit better I/O parallelism and I guess ideally there'd only be one strategy for both cases.